[Feature:InstructorUI] Notification for self registration

subCode321 opened the pull request:
### Please check if the PR fulfills these requirements:

* [ ] Tests for the changes have been added/updated (if possible)
* [ ] Documentation has been updated/added if relevant
* [x] Screenshots are attached to Github PR if visual/UI changes were made

### What is the current behavior?
<!-- List issue if it fixes/closes/implements one using the "Fixes #<number>" or "Closes #<number>" syntax -->
When a student self registers for a course, an email notification is sent to all instructors by default

### What is the new behavior?
Fixes #11317 

New option enabled to **opt out** from email notifications if a course instructor wants to!

Also, to opt-in or opt-out for `notifications` is disabled as currently we don't send notifications and only emails

![image](https://github.com/user-attachments/assets/67979464-7f47-4830-8585-029472841519)

### Other information?
<!-- Is this a breaking change? -->
<!-- How did you test -->

subCode321 commented:
hi @bmcutler , can I please get a review on this PR? Thank you 😃

subCode321 commented on site/app/controllers/course/CourseRegistrationController.php:
```diff
@@ -16,18 +17,33 @@ private function notifyInstructors(string $user, string $term, string $course, a
-                    "subject" => $subject,
-                    "body" => $body,
-                    "to_user_id" => $instructor
-                ]
-            );
+            // Find the row in the settings for the instructor
+            $instructor_settings_row = array_values(array_filter($instructors_settings, function ($v) use ($instructor) {
+                return $v['user_id'] === $instructor;
+            }));
```
Instead of running one `for loop` outside and then filtering ... I'll squash everything into one single `for loop`

IDzyre commented on site/app/controllers/course/CourseRegistrationController.php:
```diff
@@ -16,18 +17,33 @@ private function notifyInstructors(string $user, string $term, string $course, a
         }
 
-        $this->core->getNotificationFactory()->sendEmails($emails);
+        if (!empty($emails)) {
```
The sendEmails function already handles empty arrays, so I don't know if this is necessary

subCode321 replied:
I see, will change this then

IDzyre commented on site/app/controllers/NotificationController.php:
```diff
@@ -36,7 +36,8 @@ class NotificationController extends AbstractController {
         'team_member_submission_email',
-        'self_notification_email'
+        'self_notification_email',
+        'all_new_self_registrations_email'
```
Would this better fit with the other names as something like `self_registration_email` or something similar? Not a required change, just thinking about consistency.

subCode321 replied:
Ok, will keep it consistent

IDzyre commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,24 @@
+                        <span class="option-alt">Alert me when a student self registers for this course</span>
+                    </span>
+                </div>
+                <div class="option-input col-md-3 notification-setting-input"><input type="checkbox" aria-label="Notification All New Self Registrations" name="all_new_self_registrations" id="all_new_self_registrations" disabled></div>
```
Should instructors be allowed to disable on site notifications for this as well? If not, you should add the `checked` keyword to make this consistent with the other rows where it is disabled. 

Also, if you could split this line, and the ones below to make them easier to read.

subCode321 replied:
Currently, we don't send notifications for self registrations to instructors as per the logic, so that's the reason why I kept it disabled with the checkbox being `unchecked`

IDzyre commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,24 @@
+                </div>
+                <div class="option-input col-md-3 notification-setting-input"><input type="checkbox" aria-label="Notification All New Self Registrations" name="all_new_self_registrations" id="all_new_self_registrations" disabled></div>
+                {% if email_enabled %}
+                    <div class="option-input col-md-5 email-setting-input"><input type="checkbox" aria-label="Email All New Self Registrations" name="all_new_self_registrations_email" id="all_new_self_registrations_email" {% if notification_saves['all_new_self_registrations_email'] %}checked="checked"{% endif %}></div>
```
You don't need to have `checked="checked"`, you should just put `checked`.

IDzyre reviewed (changes requested):
A few style changes, and other code changes.

subCode321 commented on site/app/controllers/course/CourseRegistrationController.php:
```diff
@@ -16,18 +17,33 @@ private function notifyInstructors(string $user, string $term, string $course, a
+            if (!empty($instructor_settings_row)) {
+                $instructor_settings_row = $instructor_settings_row[0];
+            }
+            $notification_settings = User::constructNotificationSettings($instructor_settings_row);
```
Yeah, it makes sense to do it inline instead of constructing the settings

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,24 @@
+                        <span class="option-alt">Alert me when a student self registers for this course</span>
+                    </span>
+                </div>
+                <div class="option-input col-md-3 notification-setting-input"><input type="checkbox" aria-label="Notification All New Self Registrations" name="self_registration" id="self_registration" disabled></div>
+                {% if email_enabled %}
+                    <div class="option-input col-md-5 email-setting-input"><input type="checkbox" aria-label="Email All New Self Registrations" name="self_registration_email" id="self_registration_email" {% if notification_saves['self_registration_email'] %} checked {% endif %}></div>
```
These lines are pretty long.  Can you wrap them to keep them within a reasonable screen width?

subCode321 replied:
Done

williamjallen commented on site/tests/app/models/UserTester.php:
```diff
@@ -192,7 +192,8 @@ public function testToObject() {
                 'team_member_submission_email' => true,
-                'self_notification_email' => false
+                'self_notification_email' => false,
+                'self_registration_email' => true
```
Let's add a trailing comma here to make future changes cleaner in the diff:
```suggestion
                'self_registration_email' => true,
```

subCode321 replied:
Done

williamjallen commented on site/app/controllers/course/CourseRegistrationController.php:
```diff
@@ -16,15 +16,21 @@ private function notifyInstructors(string $user, string $term, string $course, a
+                    [
+                        "subject" => $subject,
+                        "body" => $body,
+                        "to_user_id" => $instructor
```
Let's also add a trailing comma here:
```suggestion
                        "to_user_id" => $instructor,
```

subCode321 replied:
Done

williamjallen reviewed (changes requested):
The code looks pretty good overall, except for a couple nitpicks.  I did not test manually, so it would be good to get a functionality review as well.

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,44 @@
+                        <span class="option-alt">Alert me when a student self registers for this course</span>
+                    </span>
+                </div>
+                <div 
+                    class="option-input 
+                    col-md-3 
+                    notification-setting-input">
```
While we don't have any code formatter for Twig code currently, I think something like this would be a bit more clear.  This is what most of the HTML formatters I've seen tend to do.  Apply below as well.

```suggestion
                <div 
                    class="option-input 
                    col-md-3 
                    notification-setting-input"
                >
```

williamjallen commented on site/app/controllers/NotificationController.php:
```diff
@@ -36,7 +36,8 @@ class NotificationController extends AbstractController {
         'team_member_submission_email',
-        'self_notification_email'
+        'self_notification_email',
+        'self_registration_email'
```
One more trailing comma to add:
```suggestion
        'self_registration_email',
```

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,44 @@
+                                type="checkbox" 
+                                aria-label="Email All New Self Registrations" 
+                                name="self_registration_email" 
+                                id="self_registration_email" 
```
As per the [style guide](https://submitty.org/developer/software_and_system_design/coding_style_guide/css#styling), IDs and classes should be dash-separated whenever possible.

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,44 @@
+            <div class="option row">
+                <div class="option-desc col-md-4">
+                    <span>
+                        <span class="option-title">New Self Registration</span>
+                        <br>
+                        <span class="option-alt">Alert me when a student self registers for this course</span>
```
I can't think of any good alternate suggestions off the top of my head, but this language feels awkward.  Can you find a more eloquent way to phrase it?  The same applies below as well.

subCode321 replied:
> I can't think of any good alternate suggestions off the top of my head, but this language feels awkward. Can you find a more eloquent way to phrase it? The same applies below as well.

done

subCode321 commented:
@williamjallen do the changes look good to go now?

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,48 @@
+                    </span>
+                </div>
+                <div 
+                    class="option-input 
+                    col-md-3 
+                    notification-setting-input"
```
Class names don't have to be wrapped.
```suggestion
                    class="option-input col-md-3 notification-setting-input"
```

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,48 @@
+                </div>
+                {% if email_enabled %}
+                    <div 
+                        class="option-input 
+                        col-md-5 
+                        email-setting-input"
```
Same here:
```suggestion
                        class="option-input col-md-5 email-setting-input"
```

williamjallen reviewed (commented):
The code looks much better now.  This still needs a functionality review by another developer, as well as a UI review from @bmcutler before this gets merged.

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,44 @@
+                            type="checkbox" 
+                            aria-label="Notification All New Self Registrations" 
+                            name="self_registration" 
+                            id="self_registration" 
```
This ID still needs to be dash-separated as per the style guide.

subCode321 replied:
sure

williamjallen replied:
@subCode321 it looks like this change still needs to be applied.

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -42,6 +42,44 @@
+            <div class="option row">
+                <div class="option-desc col-md-4">
+                    <span>
+                        <span class="option-title">New Self Registration</span>
+                        <br>
+                        <span class="option-alt">Notify me of self-registrations</span>
```
I'm still not entirely satisfied with the wording here, but I don't have a good alternative to suggest.  Deferring to @bmcutler for UI/wording review.

subCode321 commented:
Is this feature good to go now?

williamjallen commented:
@subCode321 This still needs a functionality review from another contributor plus a UI review from @bmcutler.

subCode321 commented:
> @subCode321 This still needs a functionality review from another contributor plus a UI review from @bmcutler.

Sure, let me know if anything more is needed from my side. Thanks!

IDzyre commented on site/app/templates/NotificationSettings.twig:
Since self registration is not a feature all instructors will use, the notification settings should be at the bottom of the notifications section.

subCode321 replied:
Ok, I'll add it to the bottom of the page

williamjallen replied:
Should this setting only appear if self registration is enabled perhaps?

subCode321 replied:
Yeah makes sense to only show this `new section` if self registration is enabled for the course. I'll modify

subCode321 replied:
Done and tested, the section now only appears when the self registration is enabled

subCode321 commented:
I've moved the `new section` to the bottom of the page

![image](https://github.com/user-attachments/assets/37179107-4768-4f38-991b-9e0d319f59d9)

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -287,7 +287,46 @@
+                                type="checkbox" 
+                                aria-label="Notification All New Self Registrations" 
+                                name="self_registration" 
+                                id="self_registration" 
```
```suggestion
                                id="self-registration" 
```

williamjallen commented on site/app/views/NotificationView.php:
```diff
@@ -29,7 +29,8 @@ public function showNotificationSettings($notification_saves) {
             'defaults' => User::constructNotificationSettings([]),
-            'update_settings_url' => $this->core->buildCourseUrl(['notifications', 'settings'])
+            'update_settings_url' => $this->core->buildCourseUrl(['notifications', 'settings']),
+            'self_registration_type' => $this->core->getQueries()->getSelfRegistrationType($this->core->getConfig()->getTerm(), $this->core->getConfig()->getCourse())
```
We shouldn't be executing database queries in a view.  Instead, it would be better to execute the query in the controller and pass the result to the view.

subCode321 replied:
Sure

subCode321 replied:
Done

williamjallen commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -287,7 +287,46 @@
             </div>
-
+        
+            {% if self_registration_type == constant('app\\controllers\\admin\\ConfigurationController::ALL_SELF_REGISTER') %}
```
Shouldn't this only appear for instructors?

subCode321 replied:
Checking

subCode321 replied:
Done

williamjallen reviewed (changes requested):
Please also add a Cypress test to verify that the setting only shows up if self registration is enabled for a course.

subCode321 commented:
Done, made all the required changes  and now the new section is only accessible when self registration is available, visible only to instructors and cypress tests are added

IDzyre commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -287,7 +287,46 @@
             </div>
-
+        
+            {% if core.getUser().accessFullGrading() and self_registration_type == constant('app\\controllers\\admin\\ConfigurationController::ALL_SELF_REGISTER') %}
```
`core.getUser().accessFullGrading()` should be passed in as a variable, instead of using function inside the twig.

subCode321 replied:
ok

subCode321 replied:
Done

IDzyre commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -18,6 +18,34 @@ describe('Tests for self registering for courses', () => {
+        cy.get('#self-registration').should('exist');
+
+        // restore to previous state
+        cy.visit(['testing', 'config']);
```
The testing course defaults to having self-registration enabled

subCode321 replied:
Yes but in the `beforeEach` (old code) it's unchecked before testing starts

subCode321 replied:
and I restore it back to `unchecked` so that the tests below can pass

IDzyre replied:
The next test unchecks it also, so I don't think this part is required.

subCode321 replied:
fixed

subCode321 commented:
The failing cypress is flaky, it passes sometime and sometime fails

IDzyre commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -18,6 +18,34 @@ describe('Tests for self registering for courses', () => {
+        cy.get('[data-testid="default-section-id"]').select('1');
+
+        cy.visit(['testing', 'notifications', 'settings']);
+        cy.get('#self-registration').should('not.exist');
```
Is there a testid you can use for this ?

subCode321 replied:
I'll add

subCode321 replied:
added

IDzyre commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -8,14 +8,20 @@ const notifiedMessage = 'Your instructor will be notified and can then choose to
 const no_access_message = "You don't have access to this course.";
 
 describe('Tests for self registering for courses', () => {
-    before(() => {
```
The reason this is in the 'before' block, is because the tests should be able to be run no matter what the state of the VM is in. In CI, it will always be in the same state, however when running it locally, you should be able to pause the test halfway through, and be able to test it again. However, this test adds a user, so it isn't technically able to be rerun easily, but it is still good practice to have this type of code in a `before` block.

subCode321 replied:
Done

IDzyre commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -8,14 +8,20 @@ const notifiedMessage = 'Your instructor will be notified and can then choose to
 describe('Tests for self registering for courses', () => {
-    before(() => {
-        // Testing course is on by default, but want to test unchecking and re-checking.
+    it('Should check for visibility of self registration notification setting when self registration is allowed for a course', () => {
```
These two tests can be combined, having a two line test isn't efficient.

subCode321 replied:
done

jeffrey-cordero commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -287,7 +287,47 @@
+                                name="self_registration" 
+                                data-testid="self-registration"
+                                id="self-registration" 
+                                disabled
```
Why is this disabled? Shouldn't instructors be allowed to toggle between receiving site notifications?

jeffrey-cordero replied:
I now understand why it's disabled based on the backend code, but wouldn't a better approach be to not render this input at all or extend the functionality to allow site notifications?

subCode321 replied:
yes I can try and not render the input at all and have another follow up PR to extend this functionality

jeffrey-cordero commented on site/app/templates/NotificationSettings.twig:
```diff
@@ -287,7 +287,47 @@
             </div>
-
+        
+            {% if access_full_grading and self_registration_type == constant('app\\controllers\\admin\\ConfigurationController::ALL_SELF_REGISTER') %}
```
A boolean of `is_self_registration` should be passed to the twig file. This also doesn't correctly hide the section for courses that don't allow self-registration. I believe the correct comparing type should be `ConfigurationController::NO_SELF_REGISTER`?

subCode321 replied:
Done

jeffrey-cordero reviewed (changes requested):
The functionality around toggling between receiving self-registration emails seems to work for instructors with the provided changes. I still believe improvements can be made to the notification page, such as not rendering the section for regular courses. Cypress test comments should also be addressed, as they may help ensure the appended cases pass.

jeffrey-cordero reviewed (commented):
Changes look solid, but the new Cypress tests are still failing.

subCode321 commented:
> Changes look solid, but the new Cypress tests are still failing.

Fixed the cypress tests

bmcutler reviewed (changes requested):
The addition to the UI looks ok.

However, I tried to test this PR and it was broken. 
When emails are "sent" on the developer machine they should appear in both the "Email Status" page (available for instructors on the left sidebar).  And they should also appear in files in the mailboxes folder /var/local/submitty/logs/emails/mailboxes

When I was on this branch, no emails were being sent about self-registration.  So it was impossible to test if the new setting was working or not working for the instructor.

I tried to test a discussion forum announcement, and it did not send either, so I'm afraid that this PR broke email notifications completely.  (maybe I made a mistake with this quick forum testing, but this would be a disaster).

I switched back to main and it was ok, notifcation emails appeared where expected.

subCode321 commented:
> The addition to the UI looks ok.
> 
> However, I tried to test this PR and it was broken. When emails are "sent" on the developer machine they should appear in both the "Email Status" page (available for instructors on the left sidebar). And they should also appear in files in the mailboxes folder /var/local/submitty/logs/emails/mailboxes
> 
> When I was on this branch, no emails were being sent about self-registration. So it was impossible to test if the new setting was working or not working for the instructor.
> 
> I tried to test a discussion forum announcement, and it did not send either, so I'm afraid that this PR broke email notifications completely. (maybe I made a mistake with this quick forum testing, but this would be a disaster).
> 
> I switched back to main and it was ok, notifcation emails appeared where expected.

I'm taking a look. Things were working fine in my last push except the pull I took from main. I'll test it out and attach a screen-recording to the description

subCode321 commented:
> The addition to the UI looks ok.
> 
> However, I tried to test this PR and it was broken. When emails are "sent" on the developer machine they should appear in both the "Email Status" page (available for instructors on the left sidebar). And they should also appear in files in the mailboxes folder /var/local/submitty/logs/emails/mailboxes
> 
> When I was on this branch, no emails were being sent about self-registration. So it was impossible to test if the new setting was working or not working for the instructor.
> 
> I tried to test a discussion forum announcement, and it did not send either, so I'm afraid that this PR broke email notifications completely. (maybe I made a mistake with this quick forum testing, but this would be a disaster).
> 
> I switched back to main and it was ok, notifcation emails appeared where expected.

Hi @bmcutler, I've fixed the issue for notifications and verified that they appear in both email status and mailboxes
![image](https://github.com/user-attachments/assets/0d60670c-e3ae-4c88-b871-96c8d0030c0a)

![image](https://github.com/user-attachments/assets/127b52e7-cfbf-4cc0-b372-195ea5f8a049)

subCode321 commented:
Do the changes look good to go? I've tested the changes and they're working for me

jeffrey-cordero commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -8,13 +8,26 @@ const notifiedMessage = 'Your instructor will be notified and can then choose to
+        cy.visit(['testing', 'notifications', 'settings']);
+        cy.get('[data-testid="self-registration"]').should('not.exist');
+
+        // enable self registration
```
You should disable self-registration at the end to ensure this passes on multiple attempts. If we run this test twice, `self-registration` will exist as self-registration was enabled in the prior test.

subCode321 replied:
ok

jeffrey-cordero reviewed (approved):
The overall changes ensure the emails to instructors based on their notification preference. I verified the expected release or suppression of the notification on the email status page and the logging directory, as shown below, where other notification-related events, such as discussion forum events remained intact. I will speak with Professor Cutler this week about her experience with your PR, which may have been caused by a potential notification misconfiguration.



```bash
-rw-r-----   1 submitty_daemon submitty_course_builders   511 Feb 10 08:47 1739195222.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   524 Feb 10 08:48 1739195281.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   791 Feb 10 08:50 1739195401.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   273 Feb 10 08:51 1739195461.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   771 Feb 10 08:52 1739195521.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders  1036 Feb 10 08:53 1739195581.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   284 Feb 10 08:56 1739195761.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   285 Feb 10 08:57 1739195821.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   284 Feb 10 08:59 1739195942.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   771 Feb 10 09:00 1739196001.submitty@vagrant.msg
-rw-r-----   1 submitty_daemon submitty_course_builders   437 Feb 13 08:02 1739451722.submitty@vagrant.msg
-rw-r--r--   1 root            submitty_course_builders   437 Feb 19 09:56 1739976962.submitty@vagrant.msg
-rw-r--r--   1 root            submitty_course_builders   441 Feb 19 09:58 1739977081.submitty@vagrant.msg
-rw-r--r--   1 root            submitty_course_builders   528 Feb 19 10:08 1739977681.submitty@vagrant.msg
root@ubuntu:/usr/local/submitty# cat /var/local/submitty/logs/emails/mailboxes/instructor@example.com/1739976962.submitty@vagrant.msg
Content-Type: text/plain; charset=utf-8
TO: instructor@example.com
From: submitty@vagrant
reply-to: do-not-reply@vagrant
Subject: [Submitty blank]: Self-registration of raynod for course blank


Student raynod has self-registered for course blank for term s25.

--
NOTE: This is an automated email notification, which is unable to receive replies.
Please refer to the course syllabus for contact information for your teaching staff.



root@ubuntu:/usr/local/submitty# 
```

jeffrey-cordero commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -8,13 +8,26 @@ const notifiedMessage = 'Your instructor will be notified and can then choose to
+        cy.visit(['testing', 'notifications', 'settings']);
+        cy.get('[data-testid="self-registration"]').should('not.exist');
+
+        // enable self registration
```
I was mistaken in recommending to disable the self-registration status as the `beforeEach` already handles this, sorry. Please don't implement this logic.

subCode321 commented:
Thanks for the approval jeffrey-cordero. williamjallen, IDzyre do the changes look good to you?

jeffrey-cordero commented:
I still don't see any potential issues in other notification-related features. Announcments and setting-specific notifications, such as "all new threads," for students still work, as shown below.

![image](https://github.com/user-attachments/assets/5bfa1864-bfc9-458c-961b-655ba29ccc01)

IDzyre commented on site/cypress/e2e/Cypress-Feature/registration.spec.js:
```diff
@@ -8,13 +8,26 @@ const notifiedMessage = 'Your instructor will be notified and can then choose to
         cy.visit(['testing', 'config']);
         cy.get('[data-testid="all-self-registration"]').uncheck();
         cy.get('[data-testid="all-self-registration"]').should('not.be.checked');
-        cy.get('[data-testid="default-section-id"]').select('1');
```
Is there a reason to remove this ?

subCode321 replied:
yes, because I've created this new `beforeEach` block where I just make the instructor login and uncheck registration ... so there's no need to select a section in this uncheck flow

zacharymnp reviewed (approved):
Email notification for self-registration appears in Email Statuses and Mailboxes when the notification setting is toggled on, and it does not appear when the notification setting is toggled off.

The notification setting is only visible when the course has self-registration enabled.

williamjallen reviewed (approved):
Code looks good to me
